chain-signatures/node/src/storage/presignature_storage.rs

```diff
@@ -0,0 +1,134 @@
+use std::sync::Arc;
+
+use anyhow::Ok;
+use redis::{Commands, Connection, FromRedisValue, RedisWrite, ToRedisArgs};
+use tokio::sync::RwLock;
+use url::Url;
+
+use crate::protocol::presignature::{Presignature, PresignatureId};
+
+type PresigResult<T> = std::result::Result<T, anyhow::Error>;
+pub type PresignatureStorageBox = Box<dyn PresignatureStorage + Send + Sync>;
+pub type LockPresignatureStorageBox = Arc<RwLock<PresignatureStorageBox>>;
+
+const PRESIGNATURES_MAP_NAME: &str = "presignatures";
+const MINE_SET_NAME: &str = "presignatures_mine";
+
+pub fn init(redis_url: Url) -> PresignatureStorageBox {
+    Box::new(RedisPresignatureStorage::new(redis_url)) as PresignatureStorageBox
+}
+
+pub trait PresignatureStorage {
+    fn insert(&mut self, presignature: Presignature) -> PresigResult<()>;
+    fn insert_mine(&mut self, presignature: Presignature) -> PresigResult<()>;
+    fn contains(&mut self, id: &PresignatureId) -> PresigResult<bool>;
+    fn contains_mine(&mut self, id: &PresignatureId) -> PresigResult<bool>;
+    fn take(&mut self, id: &PresignatureId) -> PresigResult<Option<Presignature>>;
+    fn take_mine(&mut self) -> PresigResult<Option<Presignature>>;
+    fn count_all(&mut self) -> PresigResult<usize>;
+    fn count_mine(&mut self) -> PresigResult<usize>;
+}
```

Comment:
we don't need a trait at all because we use soley use Redis and have no in-memory alternative

Reply:
Ok
And we do need Arc now